### ISSUE TITLE
Terraform Plan Continually Proposes USAGE Privilege Grant Despite Successful Apply on Future File Formats

### ISSUE THREAD

**Issue by jacobcbeaudin** [Snowflake-Labs/terraform-provider-snowflake#1981]
**Provider Version**

The provider version you are using.

**Terraform Version**

terraform  = 1.4.6

snowflake = 0.68.2

The version of Terraform you were using when the bug was encountered.

**Describe the bug**

I am encountering an issue with the Snowflake Terraform provider where the terraform plan continually proposes to grant the USAGE privilege to a role, even after the plan has been successfully applied.

Here's the scenario:
1. I run terraform plan and it generates the following execution plan:
```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the
following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # module.snowflake.snowflake_grant_privileges_to_role.file_formats will be updated in-place
  ~ resource "snowflake_grant_privileges_to_role" "file_formats" {
        id                = "READ_ROLE|USAGE|false|false|false|false|false|true|false|true|||FILE FORMATS|false||true|DB"
      ~ privileges        = [
          + "USAGE",
        ]
        # (4 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }
```
2. I then run terraform apply and it appears to apply the plan successfully, indicating that the USAGE privilege will be granted to the role.

3. However, when I run terraform plan again, it still proposes to grant the USAGE privilege to the role, as if the previous apply did not take effect.

4. I have confirmed that the grant privileges to role exists in my Terraform state, but it seems that the USAGE privilege that I specify is not being applied.

**Expected behavior**

This behavior is unexpected and leads to confusion as the terraform plan output does not accurately reflect the state of the resources after terraform apply is run. I would appreciate any guidance or fixes for this issue. The role should have `USGAE` privileges on the resource specified.

**Code samples and commands**

```
locals {
  functional_roles = {
    file_format_access = {
      schema_objects = [
        {
          database_name      = snowflake_database.db.name
          privileges         = ["USAGE"]
          grant_type         = "future"
          object_type_plural = "FILE FORMATS"
        }
      ]
    }
  }
}

locals {
  role_schema_object_privileges = flatten([
    for role_name, role_info in local.functional_roles : [
      for schema_object in lookup(role_info, "schema_objects", []) : {
        role_name          = role_name
        database_name      = schema_object.database_name
        privileges         = schema_object.privileges
        grant_type         = schema_object.grant_type
        object_type_plural = schema_object.object_type_plural
      }
    ]
  ])
}

terraform {

  required_version = ">= 1.4.6"

  required_providers {

    aws = {
      source  = "hashicorp/aws"
      version = "~> 5.7"
    }

    snowflake = {
      source  = "Snowflake-Labs/snowflake"
      version = "~> 0.68"
    }
  }
}

provider "snowflake" {
  alias = "security_admin"
  role  = "SECURITYADMIN"
}

resource "snowflake_role" "functional_role" {
  provider = snowflake.security_admin
  for_each = local.functional_roles
  name     = upper(each.key)
}

resource "snowflake_database" "db" {
  provider = snowflake.sys_admin
  name     = "DB"
}

resource "snowflake_grant_privileges_to_role" "schema_objects" {
  provider = snowflake.security_admin
  for_each = { for schema_object in local.role_schema_object_privileges : "${schema_object.role_name}-ON-${schema_object.grant_type}-${replace(schema_object.object_type_plural, " ", "_")}-IN-${schema_object.database_name}" => schema_object }
  privileges = each.value.privileges
  role_name  = each.value.role_name

  dynamic "on_schema_object" {
    for_each = each.value.grant_type == "future" ? [1] : []
    content {
      # does not update for FILE FORMATS future
      future {
        object_type_plural = each.value.object_type_plural
        in_database        = each.value.database_name
      }
    }
  }

  depends_on = [
    snowflake_role.functional_role,
    snowflake_database.db
  ]
```

**Comment by jacobcbeaudin** [comment]
Update: After upgrading to Snowflake Terraform version 0.69, the issue has resolved itself.

**Comment by wh-k-klemens** [comment]
We still see this behaviour at `0.70.0`. See also here https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/1998

**Comment by attrivivekFF** [comment]
Any update on this issue. Our terraform apply keep getting failed because of this issue. I mentioned it [here](https://github.com/Snowflake-Labs/terraform-provider-snowflake/issues/1998#issuecomment-1706649708) as well.

**Comment by jacobcbeaudin** [comment]
Update: the issue is now showing up when I am granting `"ALL PRIVILEGES"`. This is happening on `account_objects`, `schemas`, specific and plural `schema_objects`. I am granting privileges dynamically with the code below
```
resource "snowflake_grant_privileges_to_role" "account_object" {
  provider   = snowflake.security_admin
  for_each   = { for account_object in local.role_account_object_privileges : "${account_object.role_name}-ON-${account_object.account_object_type}-${account_object.account_object_name}" => account_object }
  privileges = each.value.privileges
  role_name  = each.value.role_name

  on_account_object {
    object_type = each.value.account_object_type
    object_name = each.value.account_object_name
  }
}

resource "snowflake_grant_privileges_to_role" "schema" {
  provider = snowflake.security_admin
  for_each = { for schema in local.role_schema_privileges : "${schema.role_name}-ON-${schema.grant_type}-${schema.database_name}.${schema.schema_name}" => schema }

  privileges = each.value.privileges
  role_name  = each.value.role_name

  dynamic "on_schema" {
    for_each = each.value.grant_type == "all" ? [1] : []
    content {
      all_schemas_in_database = each.value.database_name
    }
  }

  dynamic "on_schema" {
    for_each = each.value.grant_type == "future" ? [1] : []
    content {
      future_schemas_in_database = each.value.database_name
    }
  }

  dynamic "on_schema" {
    for_each = each.value.grant_type == "specific" ? [1] : []
    content {
      schema_name = "\"${each.value.database_name}\".\"${each.value.schema_name}\""
    }
  }
}

resource "snowflake_grant_privileges_to_role" "schema_object" {
  provider   = snowflake.security_admin
  for_each   = { for schema_object in local.role_schema_object_privileges : "${schema_object.role_name}-ON-${replace(schema_object.object_type, " ", "_")}-${schema_object.database_name}.${schema_object.schema_name}.${schema_object.object_name}" => schema_object }
  privileges = each.value.privileges
  role_name  = each.value.role_name

  on_schema_object {
    object_type = each.value.object_type
    object_name = "\"${each.value.database_name}\".\"${each.value.schema_name}\".\"${each.value.object_name}\""
  }
}

resource "snowflake_grant_privileges_to_role" "schema_object_plural" {
  provider   = snowflake.security_admin
  for_each   = { for schema_object in local.role_schema_object_plural_privileges : "${schema_object.role_name}-ON-${schema_object.grant_type}-${replace(schema_object.object_type_plural, " ", "_")}-IN-${schema_object.database_name}.${schema_object.schema_name}" => schema_object }
  privileges = each.value.privileges
  role_name  = each.value.role_name

  dynamic "on_schema_object" {
    for_each = each.value.grant_type == "all" ? [1] : []
    content {
      all {
        object_type_plural = each.value.object_type_plural
        in_schema          = "\"${each.value.database_name}\".\"${each.value.schema_name}\""
      }
    }
  }

  dynamic "on_schema_object" {
    for_each = each.value.grant_type == "future" ? [1] : []
    content {
      future {
        object_type_plural = each.value.object_type_plural
        in_schema          = "\"${each.value.database_name}\".\"${each.value.schema_name}\""
      }
    }
  }
}
```

**Comment by sfc-gh-swinkler** [comment]
This has been fixed as of 0.70.1. I was able to verify the issue in 0.68.2, but after upgrading it is not longer an issue.

```
resource "snowflake_role" "read_role" {
  name = "READ_ROLE"
}

resource "snowflake_database" "db" {
  name = "DB"
}

resource "snowflake_grant_privileges_to_role" "file_formats" {
  privileges = ["USAGE"]
  on_schema_object {
    future {
      object_type_plural = "FILE FORMATS"
      in_database = "DB"
    }
  }
  role_name = snowflake_role.read_role.name
}
```

in 0.68.2 I get the continual privilege add

```
  # snowflake_grant_privileges_to_role.file_formats will be updated in-place
  ~ resource "snowflake_grant_privileges_to_role" "file_formats" {
        id                = "READ_ROLE|USAGE|false|false|false|false|false|true|false|true|||FILE FORMATS|false||true|DB"
      ~ privileges        = [
          + "USAGE",
        ]
        # (4 unchanged attributes hidden)

        # (1 unchanged block hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```

but after upgrading:

```
No changes. Your infrastructure matches the configuration.

Terraform has compared your real infrastructure against your configuration and
found no differences, so no changes are needed.

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.
```

If you have any further issues, please open a new ticket.
